improvements in wake word detection and truncation for better processing

HaroonAbbas12 opened the pull request:
## Summary
This pull request enhances the voice assistant by introducing wake-word functionality and improving token management for the ChatGPT API interactions.

## Changes Made
- **Wake Word Detection**: Implemented a new method to wait for the specified wake word ("Hey Harry") before activating the assistant.
- **Listening Timeout**: Added a timeout feature to the listening function to prevent indefinite waiting.
- **Token Management**: Integrated a token counting mechanism using the tiktoken library to ensure input and response token limits are adhered to.
- **Response Limitation**: Configured the assistant to provide concise responses limited to 40 words.

## Why
These enhancements aim to improve user experience by allowing the assistant to respond only upon hearing the wake word and ensuring that interactions with the ChatGPT API remain efficient and manageable.

## How
- Added `wait_for_wake_word` method to the `VoiceInput` class for wake word detection.
- Integrated the tiktoken library for calculating token counts in the `GPTProcessor` class.
- Updated the `process` method to truncate user input and responses based on token limits.

HaroonAbbas12 committed:
new updates

awaiskaleem commented on wake_word.py:
```diff
@@ -3,24 +3,31 @@
 openai.api_key = os.getenv('OPENAI_API_KEY')
 
+# Tokenizer setup for total tokens used calculation
+tokenizer = tiktoken.get_encoding("gpt2")
```
Why gpt2? Arent' we using gpt4omini?

HaroonAbbas12 replied:
https://github.com/openai/tiktoken/blob/main/tiktoken_ext/openai_public.py
because gpt 40-mini is not working not even gpt 3.5 turbo .it says that the tokenizer can not recognize that .i have added a link from where i have added a gpt 2

awaiskaleem replied:
Maybe you should use `cl100k_base` from [here](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_count_tokens_with_tiktoken.ipynb)?

HaroonAbbas12 replied:
yes i used that first but gpt 2 was also working so i added that, what is your suggestion ?

HaroonAbbas12 replied:
alright?i was not aware of the differences they create .**cl100k_base** is more advanced when you are using gpt 4 models ,i will change it.

awaiskaleem commented on wake_word.py:
```diff
@@ -3,24 +3,31 @@
         self.microphone = sr.Microphone()
 
-    def listen(self):
+    def listen(self, timeout=5):  # Added timeout parameter
```
Kindly change this to 10.

HaroonAbbas12 replied:
okay

awaiskaleem commented on wake_word.py:
```diff
@@ -33,26 +40,41 @@ def listen(self):
             return None
 
-    def wait_for_wake_word(self, wake_word="Hey Harry"):
+    def wait_for_wake_word(self, wake_word="Hey Harry", max_attempts=3):
```
What does this `max_attemps` do? Can you explain here in this comment?

HaroonAbbas12 replied:
If the wake word is not detected with in 3 attempts it will exit the code ..

awaiskaleem replied:
What is your thinking behind it. Ideally, it should just go back to "sleep" where it can listen to wake word again right? If the code exits, it will be difficult to start it again on a RaspberryPi.

HaroonAbbas12 replied:
yes,if we start the program again and again it is not appropriate as a user experience .will work on that also.

awaiskaleem commented on wake_word.py:
```diff
@@ -62,12 +84,23 @@ def process(self, text):
+
+            # Token calculation for response
+            response_tokens = self.count_tokens(message_content)
+            if response_tokens > 40:  # Ensure response tokens do not exceed 40
```
Do you think we should not put a limit to response? OpenAI is going to charge us anyway, why not receive full response. Notice that you control the `temperature` of response while making the call to OpenAI

HaroonAbbas12 replied:
we have a limit of around 100 tokens per minute so if we change the output response it will cost us more because the number of output tokens will increase .and tokens are dependant on the words you are using .special characters use more tokens then simple words.

awaiskaleem replied:
I understand but I think we are mixing "input tokens" (what we send to OpenAI) and "response tokens" (what OpenAI send us in response to our prompt). We shouldn't have any limit on response, or a very big limit on response, like 1000 or something.

awaiskaleem reviewed (changes requested):
Left some comments.

awaiskaleem commented:
Good job with writing the PR summary.

HaroonAbbas12 commented:
Thank you .

HaroonAbbas12 committed:
Added input truncation and hibernation mode

- Implemented input truncation to limit user input to 100  words for optimized processing.
- Introduced hibernation mode: the system re-enters listening state after detecting the wake word

HaroonAbbas12 commented:
I've made some additional changes to the input truncation and hibernation mode logic. Can you please review the latest updates?

awaiskaleem commented:
LGTM (Looks good to me). Please merge this and send a video of how this works right now.

awaiskaleem reviewed (approved):
LGTM
